### ISSUE TITLE
Question about the loss computation

### ISSUE THREAD

**Issue by zhaoyu-li** [LUMII-Syslab/QuerySAT#1]
Hi, 

I read your paper and find that it is very interesting. However, I am a little confused with your implementation of the clause loss `V_c(x)` stated in your paper, which are the functions `softplus_mixed_loss` and `softplus_loss`.

Since we need to compute `log V_c(x) = log(1-c_loss)` for each clause and there are some multiplication computations to compute c_loss. It seems like you do some transformations to move the log to c_loss and transform multiplication to addition. Could you please explain why you add exponential in the function `softplus_loss` and what `log_clauses` mean in the function `softplus_mixed_loss`?

Thank you in advance!

**Comment by EmilsOzolins** [comment]
Hi!

We are glad that you found our paper interesting. 

Your defined `c_loss` corresponds to `1-V_c` in our paper and `softplus_loss` in our code. To squash model output logits to the range [0,1], we use the sigmoid function. Unfortunately, if we use sigmoid and multiplication to calculate `c_loss`, then numerical instabilities could cause results to tend towards zero even when they should be one.`Softplus_loss` calculates the same loss as sigmoid with multiplication but avoids numerical instabilities as instead of multiplication we perform addition in log-space.

`Softplus_mixed_loss` on the other hand corresponds to the `log(1-c_loss)`. As the output of the `softplus_loss` is linear space, we still need to move to the log-space. Additionally, we multiply log-space loss with the linear-space loss for two reasons. Firstly, it offsets the square root that we take for each graph latter in loss calculations as the `log(1-c_loss)*c_loss` serves as the equivalent to `log(1-c_loss)^2`. Secondly, we find that on some tasks, mixed version works slightly better than squared log-loss.

I hope that this answers your questions, and let us know if anything else is unclear.

**Comment by zhaoyu-li** [comment]
Got it! Thank you for your quick reply! By the way, there is an interesting question: is there any connection between your defined loss and using MAX to replace AND and using MIN to replace OR? I just think about it. Their values are the same when the assignments are hard, however, when the assignments are soft, their solution space is quite different.

**Comment by EmilsOzolins** [comment]
One could indeed formulate a similar loss using softmax and softmin functions for approximating OR and AND, respectively. Nonetheless, training the network using such loss requires introducing a temperature parameter that is gradually annealed to achieve max and min functions. Amizadeh et al. (2019) and Amizadeh et al. (2019) have explored such an option to solve Circuit-SAT and SAT. In our experiments, softmax/softmin loss converged much slower and never achieved the same accuracy as our loss. It was also cumbersome to find a valid annealing rate for temperature parameters.

**Comment by zhaoyu-li** [comment]
Yeah, thank you! 

Sorry, I still think about your implementation of `linear_loss` and `softplus_loss`and there are still some problems I am confused with.

In line 25 & 39, `clauses_val = tf.sparse.sparse_dense_matmul(adj_matrix, literals)` you actually use matrix multiplication and the `adj_matrix` is of the shape [n_literals, n_clauses] and the `literals` is of the shape [n_literals, 1] and the `clauses_val` is of the shape [n_clauses, 1], what you actually do in this line is for each clause, sum up the literal assignments, not multiply the assignment as stated in your paper, is there some misunderstanding? 

Thanks again for your quick reply!

**Comment by EmilsOzolins** [comment]
Linear_loss is not directly related to our paper but somehow just crept into the public repository, we have deleted it. Sorry if that has caused any misunderstanding.

Regarding matrix multiplication in softplus_loss - we sum up softplus of literals which is equal to the sum of -log(sigmoid(-l_i)), where l_i is i-th literal in clause. The result of softplus_loss therefore is equal to multiplying literals in each clause together. For example, if we have an SAT formula with only one clause and n literals, then it is easy to see that the result of softplus_loss is equal to the formulas stated in our paper (see formulas below). In the formulas, the i-th literal is devoted as l_i, and the sigma depicts the sigmoid function, x_i and c+ and c- coincides with paper definitions.

![image](https://user-images.githubusercontent.com/16975996/139406474-b6d20702-e551-4db8-8db4-c519870552ea.png)

**Comment by zhaoyu-li** [comment]
Thank you very much!
